src/main/java/com/avispl/symphony/dal/avdevices/microphone/sennheiser/tcc2/SennheiserTCC2Communicator.java

```diff
@@ -0,0 +1,846 @@
+			reentrantLock.unlock();
+		}
+		return Collections.singletonList(localExtendedStatistics);
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	protected void internalInit() throws Exception {
+		fetchingDataExSer = Executors.newFixedThreadPool(1);
+		timeoutManagementExSer = Executors.newFixedThreadPool(1);
+		super.internalInit();
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	protected void internalDestroy() {
+		countMonitoringAndControllingCommand = 0;
+		if (localExtendedStatistics != null && localExtendedStatistics.getStatistics() != null && localExtendedStatistics.getControllableProperties() != null) {
+			localExtendedStatistics.getStatistics().clear();
+			localExtendedStatistics.getControllableProperties().clear();
+		}
+		if (!localCacheMapOfPropertyNameAndValue.isEmpty()) {
+			localCacheMapOfPropertyNameAndValue.clear();
+		}
+		isEmergencyDelivery = false;
+		isConfigManagement = false;
+		failedMonitor.clear();
+		try {
+			fetchingDataExSer.shutdownNow();
+			timeoutManagementExSer.shutdownNow();
+		} catch (Exception e) {
+			logger.warn("Unable to end the TCP connection.", e);
+		} finally {
+			super.internalDestroy();
+		}
+	}
+
+	/**
+	 * Using multi thread to implement get request
+	 * Thread 1 retrieves data.
+	 * Thread 2 manage the request timeout of thread 1
+	 */
+	private void retrieveMonitoringAndControllingData() {
+		List<SennheiserPropertiesList> commands = Arrays.asList(SennheiserPropertiesList.values());
+		for (int i = 0; i < commands.size(); i++) {
+			SennheiserPropertiesList commandIndex = commands.get(i);
+			if (!isConfigManagement && commandIndex.isControl()) {
+				continue;
+			}
+			// submit a thread to fetch data from the device.
+			devicesExecutionPool.add(fetchingDataExSer.submit(() -> {
+				retrieveDataByCommandName(commandIndex);
+			}));
+
+			//Using 2nd thread to monitor timeouts for commands executed in thread 1
+			Future manageTimeOutWorkerThread = timeoutManagementExSer.submit(() -> {
+				int timeoutCount = 1;
+				while (!devicesExecutionPool.get(devicesExecutionPool.size() - SennheiserConstant.ORDINAL_TO_INDEX_CONVERT_FACTOR).isDone() && timeoutCount <= SennheiserConstant.TIME_OUT_COUNT) {
+					try {
+						Thread.sleep(100);
+					} catch (InterruptedException e) {
+						e.printStackTrace();
+					}
+					timeoutCount++;
+				}
+				//If the Future is not completed after the defaultConfigTimeout =>  update the failedMonitor and destroy the connection.
+				int lastIndex = devicesExecutionPool.size() - 1;
+				if (!devicesExecutionPool.get(lastIndex).isDone()) {
+					failedMonitor.add(commandIndex.getName());
+					destroyChannel();
+					devicesExecutionPool.get(lastIndex).cancel(true);
+				}
+			});
+			try {
+				while (!manageTimeOutWorkerThread.isDone()) {
+					Thread.sleep(100);
+				}
+				manageTimeOutWorkerThread.get();
+			} catch (ExecutionException e) {
+				e.printStackTrace();
+			} catch (InterruptedException e) {
+				throw new RuntimeException(e);
```

Comment:
is there any reason to rethrow interrupted exception as Runtime here? Does it require any additional processing?